libcudacxx/include/cuda/std/__utility/cmp.h

```diff
@@ -31,99 +34,125 @@
 
 _CCCL_PUSH_MACROS
 
+_CCCL_DIAG_PUSH
+_CCCL_DIAG_SUPPRESS_MSVC(4389) // signed/unsigned mismatch
+
 _LIBCUDACXX_BEGIN_NAMESPACE_STD
 
-#if _CCCL_STD_VER > 2017
 template <class _Tp, class... _Up>
-struct _IsSameAsAny : _Or<_IsSame<_Tp, _Up>...>
-{};
+using __is_same_as_any = __fold_or<_CCCL_TRAIT(is_same, _Tp, _Up)...>;
 
 template <class _Tp>
-concept __is_safe_integral_cmp =
-  is_integral_v<_Tp>
-  && !_IsSameAsAny<_Tp,
-                   bool,
-                   char,
-                   char16_t,
-                   char32_t
-#  ifndef _LIBCUDACXX_NO_HAS_CHAR8_T
-                   ,
-                   char8_t
-#  endif
-#  ifndef _LIBCUDACXX_HAS_NO_WIDE_CHARACTERS
-                   ,
-                   wchar_t
-#  endif
-                   >::value;
-
-template <__is_safe_integral_cmp _Tp, __is_safe_integral_cmp _Up>
+struct __is_safe_integral_cmp
+    : bool_constant<_CCCL_TRAIT(is_integral, _Tp)
+                    && !__is_same_as_any<_Tp,
+                                         bool,
+                                         char,
+                                         char16_t,
+                                         char32_t
+#ifndef _LIBCUDACXX_NO_HAS_CHAR8_T
+                                         ,
+                                         char8_t
+#endif // _LIBCUDACXX_NO_HAS_CHAR8_T
+#ifndef _LIBCUDACXX_HAS_NO_WIDE_CHARACTERS
+                                         ,
+                                         wchar_t
+#endif // _LIBCUDACXX_HAS_NO_WIDE_CHARACTERS
+                                         >::value>
+{};
+
+_CCCL_TEMPLATE(class _Tp, class _Up)
+_CCCL_REQUIRES(__is_safe_integral_cmp<_Tp>::value _CCCL_AND __is_safe_integral_cmp<_Up>::value)
 _LIBCUDACXX_HIDE_FROM_ABI constexpr bool cmp_equal(_Tp __t, _Up __u) noexcept
 {
-  if constexpr (is_signed_v<_Tp> == is_signed_v<_Up>)
+#if _CCCL_STD_VER >= 2017
+  _CCCL_IF_CONSTEXPR (_CCCL_TRAIT(is_signed, _Tp) == _CCCL_TRAIT(is_signed, _Up))
```

Comment:
Q: Why do we need `_CCCL_IF_CONSTEXPR` here, if we already guard this code by `#if _CCCL_STD_VER >= 2017`?
```suggestion
#if _CCCL_STD_VER >= 2017
  if constexpr (_CCCL_TRAIT(is_signed, _Tp) == _CCCL_TRAIT(is_signed, _Up))
```
There are 3 more occurences of this below.

Reply:
Because we don't check if `__cpp_if_constexpr` is defined. Maybe we can add a `_CCCL_NO_IF_CONSTEXPR` config macro similar to `_CCCL_NO_CONCEPTS`

Reply:
Oh, it is actually defined, I didn't notice